### PR TITLE
Implement rule versioning

### LINKED ISSUE #448: Avoid annoyance from new rules
Quoting @Blacksmoke16 from https://github.com/crystal-ameba/ameba/issues/447#issuecomment-1892747306

> It just feels less than ideal that new rules keep being added that cannot be actually robust without semantic analysis, or are more subjective in general. Which ends up forcing me to either disable them inline, globally, or conform to the suggestions on every minor release.
>
> I still think some/most of these should be disabled by default, or made optional via some sort of opinionated style https://github.com/crystal-ameba/ameba/pull/112.

I think this is quite similar to the considerations about the Crystal formatter rolling out new formats (c.f. https://github.com/crystal-lang/crystal/issues/13002).
The formatter is tied to Crystal releases. Ameba on the other hand has more freedom about chosing when to release new rules.
How exactly that can be utilized, is up for discussion.

But in general, I think it would be very helpful to avoid required user interaction every time a new rule is added. That's especially if the rule has a high rate of false positives (such as `Lint/UselessAssign`). That just keeps annoying users.

Maybe for a project with existing ameba configuration, all rules that have been added since the version that generated the ameba config could be disabled by default? I.e. if I have a config from ameba 1.6, ameba will only consider rules from 1.6 until I explicitly upgrade to a new version and ruleset.

### PR DESCRIPTION
**Pull request by Sija**
- Adds `Version` property to `.ameba.yml` (optional)
- Adds `Rule::Base#since_version` (optional)
- Adds `--up-to-version <VERSION>` CLI switch
- Adds `--rule-versions` CLI switch

image

Resolves #448
Refs #470, #307

### CODE REVIEW

**Review by devnote-dev — APPROVED** [review]

**Comment by devnote-dev** [comment]
`ameba -u` working great! I have a somewhat long list of rules and the versions they were added in if you need it, it goes back as far as 1.3.0. All that remains is implementing the warning message. A compile time flag seems best here, I'm not picky about any particular name for it.

**Comment by Sija** [comment]
@devnote-dev I'll go ahead and finalize this PR, merge it and then let's move onto the next points in the following PRs.

**Review by veelenga — APPROVED** [review]
lgtm 👏🏻

File: src/ameba/presenter/rule_versions_presenter.cr [hunk 1]
```diff
@@ -0,0 +1,12 @@
+module Ameba::Presenter
+  class RuleVersionsPresenter < BasePresenter
+    def run(rules)
+      versions =
+        rules.compact_map(&.since_version).sort!.uniq!
+
+      versions.each do |version|
+        output.puts "- %s" % version.to_s.colorize(:green)
```
Comment (devnote-dev) [inline src/ameba/presenter/rule_versions_presenter.cr:8]:
```suggestion
        output << "- " << version.colorize(:green) << '\n'
```

File: src/ameba/presenter/rule_presenter.cr [hunk 2]
```diff
@@ -1,14 +1,24 @@
...
-      output.puts
       output_title "Rule info"
-      output_paragraph "%s of a %s severity [enabled: %s]" % {
+
+      info = <<-INFO
+        Name:           %s
+        Severity:       %s
+        Enabled:        %s
+        Since version:  %s
+        INFO
```
Comment (devnote-dev) [inline src/ameba/presenter/rule_presenter.cr:11]: Couldn't we make use of `String.build` here then pass it to `output_paragraph`?

File: src/ameba/config.cr [hunk 3]
```diff
@@ -100,6 +104,9 @@ class Ameba::Config
     @excluded = load_array_section(config, "Excluded")
     @globs = load_array_section(config, "Globs", DEFAULT_GLOBS)
 
+    if version = config["version"]?.try(&.as_s).presence
```
Comment (devnote-dev) [inline src/ameba/config.cr:107]:
```suggestion
    if version = config["Version"]?.try(&.as_s).presence
```

File: src/ameba/presenter/rule_presenter.cr (hunk 2 above)
Comment (Sija) [reply src/ameba/presenter/rule_presenter.cr:11]: I see no reason to do so.

**Review by devnote-dev — APPROVED** [review]

File: spec/ameba/presenter/rule_versions_presenter_spec.cr [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        versions = rules
+          .sort_by { |rule| rule.since_version || missing_version }
+          .group_by(&.since_version)
+
+        versions.each do |version, version_rules|
+          output.should contain (version || "N/A").to_s
+
+          version_rules.map(&.name).each do |name|
+            output.should contain name
+          end
```
Comment (straight-shoota) [inline spec/ameba/presenter/rule_versions_presenter_spec.cr:29]: **thought:** This is close to a re-implementation of `RuleVersionsPresenter`. If both have the same logic error, the testing is pretty useless. Also the tests do not establish any relation between the individual items, so there could be many wrong outputs that still pass the test.
Why not test against a fixed output string?

File: src/ameba/runner.cr [hunk 5]
```diff
@@ -59,18 +62,27 @@ module Ameba
...
-        config.rules
-          .find &.class.==(Rule::Lint::UnneededDisableDirective)
+        rules.find &.class.==(Rule::Lint::UnneededDisableDirective)
     end
 
-    protected def initialize(@rules, @sources, @formatter, @severity, @autocorrect = false)
+    protected def rule_runnable?(rule)
+      rule.enabled? && !rule.special? &&
+        (!(version = @version) || !(since_version = rule.since_version) ||
+          since_version <= version)
```
[the remark this answers is not in the document]
Comment (Sija) [reply src/ameba/runner.cr:85]: Agreed, although IMO it still small enough to let it be. I'll happily refactor it once it becomes unwieldy.

**Comment by devnote-dev** [comment]
Is there anything left to be added for this to be merged?

**Comment by Sija** [comment]
@devnote-dev I'm thinking of making another point-release before switching to 1.7, so until then I want to wait with merging this PR.

File: src/ameba/runner.cr (hunk 5 above)
[the remark this answers is not in the document]
Comment (Sija) [reply src/ameba/runner.cr]: Addressed in 2e62b5d0b3da2d682a0a94b63f032ca5a620d79b

**Pull request merged**